flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ComparableTypeStrategy.java

```diff
@@ -97,100 +98,84 @@ public ArgumentCount getArgumentCount() {
-		}
-
-		if (hasFamily(firstLogicalType, LogicalTypeFamily.NUMERIC) &&
-				hasFamily(secondLogicalType, LogicalTypeFamily.NUMERIC)) {
+		if (hasFamily(firstType, LogicalTypeFamily.NUMERIC) && hasFamily(secondType, LogicalTypeFamily.NUMERIC)) {
 			return true;
 		}
 
 		// DATE + ALL TIMESTAMPS
-		if (hasFamily(firstLogicalType, LogicalTypeFamily.DATETIME) &&
-				hasFamily(secondLogicalType, LogicalTypeFamily.DATETIME)) {
+		if (hasFamily(firstType, LogicalTypeFamily.DATETIME) && hasFamily(secondType, LogicalTypeFamily.DATETIME)) {
 			return true;
 		}
 
 		// VARCHAR + CHAR (we do not compare collations here)
-		if (hasFamily(firstLogicalType, LogicalTypeFamily.CHARACTER_STRING) &&
-				hasFamily(secondLogicalType, LogicalTypeFamily.CHARACTER_STRING)) {
+		if (hasFamily(firstType, LogicalTypeFamily.CHARACTER_STRING) &&
+				hasFamily(secondType, LogicalTypeFamily.CHARACTER_STRING)) {
 			return true;
 		}
 
 		// VARBINARY + BINARY
-		if (hasFamily(firstLogicalType, LogicalTypeFamily.BINARY_STRING) &&
-				hasFamily(secondLogicalType, LogicalTypeFamily.BINARY_STRING)) {
+		if (hasFamily(firstType, LogicalTypeFamily.BINARY_STRING) &&
+				hasFamily(secondType, LogicalTypeFamily.BINARY_STRING)) {
 			return true;
 		}
 
 		return false;
 	}
 
-	private boolean areRawTypesComparable(LogicalType firstLogicalType, LogicalType secondLogicalType) {
-		return firstLogicalType.copy(true).equals(secondLogicalType.copy(true)) &&
-			Comparable.class.isAssignableFrom(((RawType<?>) firstLogicalType).getOriginatingClass());
+	private boolean areTypesOfSameRootComparable(LogicalType firstType, LogicalType secondType) {
+		switch (firstType.getTypeRoot()) {
+			case ARRAY:
+			case MULTISET:
+			case MAP:
+			case ROW:
+				return areCollectionsComparable(firstType, secondType);
+			case DISTINCT_TYPE:
+				return areDistinctTypesComparable(firstType, secondType);
+			case STRUCTURED_TYPE:
+				return areStructuredTypesComparable(firstType, secondType);
+			case RAW:
+				return areRawTypesComparable(firstType, secondType);
+			default:
+				return true;
+		}
+	}
+
+	private boolean areRawTypesComparable(LogicalType firstType, LogicalType secondType) {
+		return firstType.equals(secondType) &&
+			Comparable.class.isAssignableFrom(((RawType<?>) firstType).getOriginatingClass());
 	}
 
-	private boolean areDistinctTypesComparable(LogicalType firstLogicalType, LogicalType secondLogicalType) {
-		DistinctType firstDistinctType = (DistinctType) firstLogicalType;
-		DistinctType secondDistinctType = (DistinctType) secondLogicalType;
-		return firstLogicalType.copy(true).equals(secondLogicalType.copy(true)) &&
+	private boolean areDistinctTypesComparable(LogicalType firstType, LogicalType secondType) {
+		DistinctType firstDistinctType = (DistinctType) firstType;
+		DistinctType secondDistinctType = (DistinctType) secondType;
+		return firstType.equals(secondType) &&
 			areComparable(firstDistinctType.getSourceType(), secondDistinctType.getSourceType());
```

Comment:
@dawidwys One last minute comment: We need to remove the nullability for all calls of `areComparable` here and for other types. Maybe it is better to introduce an intermediate function call.